toolkit/tools/pkg/licensecheck/licensecheck.go

```diff
@@ -0,0 +1,411 @@
+macros, i.e. a chroot): CheckRpmLicenses(), IsALicenseFile(), IsASkippedLicenseFile()
+
+The LicenseCheckerResult struct is used to store the results of the search. It contains the path to the RPM, a list of
+bad documents, a list of bad files, and a list of duplicated documents. The bad documents are %doc files that are not
+at least also in the license files. The bad files are general files that are misplaced in the licenses directory.
+
+The duplicated documents are %doc files that are also in the license files. These are not technically bad, but are messy
+and should be cleaned up.
+*/
+package licensecheck
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/logger"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/rpm"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/sliceutils"
+	"github.com/microsoft/azurelinux/toolkit/tools/pkg/simpletoolchroot"
+)
+
+const (
+	chrootName      = "license_chroot"
+	licensePrefix   = "/usr/share/licenses"
+	numParallelJobs = 20
+)
+
+// LicenseChecker is a tool for searching RPMs for bad licenses
+type LicenseChecker struct {
+	simpleToolChroot *simpletoolchroot.SimpleToolChroot // The chroot to scan the RPMs in
+	distTag          string                             // The distribution tag to use when parsing RPMs
+	licenseNames     LicenseNames                       // The regexes used to match license files
+	exceptions       LicenseExceptions                  // Files that should be ignored
+	results          []LicenseCheckResult               // The results of the search
+	jobSemaphore     chan struct{}                      // Limit the number of parallel jobs
+}
+
+// New creates a new license checker. If this returns successfully the caller is responsible for calling CleanUp().
+// - buildDirPath: The path to create the chroot inside
+// - workerTarPath: The path to the worker tarball
+// - rpmDirPath: The path to the directory containing the RPMs
+// - nameFilePath: The path to the .json file containing license names
+// - exceptionFilePath: Optional, the path to the .json file containing license exceptions to ignore
+// - distTag: The distribution tag to use when parsing RPMs
+func New(buildDirPath, workerTarPath, rpmDirPath, nameFilePath, exceptionFilePath, distTag string,
+) (newLicenseChecker *LicenseChecker, err error) {
+	newLicenseChecker = &LicenseChecker{
+		distTag:          distTag,
+		simpleToolChroot: &simpletoolchroot.SimpleToolChroot{},
+		jobSemaphore:     make(chan struct{}, numParallelJobs),
+	}
+
+	err = newLicenseChecker.simpleToolChroot.InitializeChroot(buildDirPath, chrootName, workerTarPath, rpmDirPath)
+	if err != nil {
+		newLicenseChecker.CleanUp()
+		err = fmt.Errorf("failed to initialize chroot:\n%w", err)
+		return nil, err
+	}
+
+	newLicenseChecker.licenseNames, err = LoadLicenseNames(nameFilePath)
```

Comment:
This looks a bit fishy: why is `LoadLicenseNames` a public function if it's only used internally here and in the tests? We can always make it public, if it makes sense to run on its own but until then, there's no need to expose it. Especially since I understand we want to use the OO model, where all interactions with the module are handled through the `LicenseChecker` objects.

Reply:
Its used in the test program `licensetestchecker.go`. If we really want to I can write a wrapper around it and `LoacLicenseExceptions, but I don't see the harm in having a public factory/constructor function for the `LicenseNames` and `LicenseExceptions` types.

(I had a bug in the test program that was using the old pattern to call `IsALicenseFile()`, fixed now.